runtime/integration-tests/src/pools/utils/loans.rs

```diff
@@ -0,0 +1,245 @@
+use pallet_loans::{
+	loan_type::{BulletLoan, LoanType},
+	math::interest_rate_per_sec,
+	types::Asset,
+	Call as LoansCall,
+};
+use pallet_uniques::Call as UniquesCall;
+use runtime_common::{AccountId, ClassId, InstanceId, PoolId};
+use std::collections::HashMap;
+
+/// Structure that manages collateral and loan nft ids
+pub struct NftManager {
+	collaterals: HashMap<PoolId, InstanceId>,
+	loans: HashMap<PoolId, InstanceId>,
+}
+
+/// The id we use for loans
+pub type LoanId = InstanceId;
+
+// The id we use for collaterals
+pub type CollateralId = InstanceId;
+
+impl NftManager {
+	pub fn new() -> Self {
+		Self {
+			collaterals: HashMap::new(),
+			loans: HashMap::new(),
+		}
+	}
+
+	/// Currently simply maps pool_id = loan_class_id for a pool
+	pub fn loan_class_id(&self, pool_id: PoolId) -> ClassId {
+		pool_id
+	}
+
+	/// Maps (pool_id + 1) << 32 = collateral_class id
+	///
+	/// panics if pool_id >= u32::MAX - 1 as this would result in an overflow
+	/// during shifting.
+	pub fn collateral_class_id(&self, pool_id: PoolId) -> ClassId {
+		assert!(
+			pool_id < u32::MAX.into(),
+			"Pool-id must be smaller u32::MAX for testing. To ensure no-clashes in NFT class-ids"
+		);
+		let id = (pool_id + 1) << 32;
+		id
+	}
+
+	pub fn curr_loan_id(&mut self, pool_id: PoolId) -> InstanceId {
+		self.loans.entry(pool_id).or_insert(InstanceId(1)).clone()
+	}
+
+	fn next_loan_id(&mut self, pool_id: PoolId) -> InstanceId {
+		let id = self.loans.entry(pool_id).or_insert(InstanceId(1));
+		let next = id.clone();
+		*id = InstanceId(id.0);
+		next
+	}
+
+	pub fn curr_collateral_id(&mut self, pool_id: PoolId) -> InstanceId {
+		self.loans.entry(pool_id).or_insert(InstanceId(1)).clone()
+	}
+
+	fn next_collateral_id(&mut self, pool_id: PoolId) -> InstanceId {
+		let id = self.collaterals.entry(pool_id).or_insert(InstanceId(1));
+		let next = id.clone();
+		*id = InstanceId(id.0);
+		next
+	}
+}
+
+/// Creates the necessary extrinsics to initialises a pool in the loans pallet.
+/// The pool must already exist for this extrinsics to succeed.
+///
+/// Extrinsics that are generated:
+/// * Loans::initialise_pool
+/// * Uniques::create -> for Loan nft class
+/// * Uniques::create -> for Collateral nft class
+pub fn init_loans_for_pool(
+	owner: AccountId,
+	pool_id: PoolId,
+	manager: &mut NftManager,
+) -> Vec<Call> {
+	let loan_class = manager.loan_class_id(pool_id);
+	let collateral_class = manager.collateral_class_id(pool_id);
+	let mut calls = Vec::new();
+	calls.push(create_nft_call(owner.clone(), loan_class));
+	calls.push(create_nft_call(owner, collateral_class));
+	calls.push(initialise_pool_call(pool_id, loan_class));
+	calls
+}
+
+/// Issues a default loan with the following properties
+/// * 15% APR
+/// * value with amount
+/// * maturity as given
+/// * Type: BulletLoan
+/// 	* advance_rate: 90%,
+///     * probability_of_default: 5%,
+///     * loss_given_default: 50%,
+/// 	* discount_rate: 4% ,
```

Comment:
Just an FYI that a more realistic example is that `discount_rate` is slightly below the interest rate. E.g. a discount rate of 3.5% for an interest rate of 4%. If `interest rate == discount rate`, then the debt basically grows at the same rate as the present value (if we exclude risk adjustments from PD & LGD), which isn't a practical example as future cashflows should generally be discounted at present.

Reply:
Thanks! Makes sense. And for the assumed interest rate we take the interest rate of the top-non-residual tranche (i.e. the most senior one)?

Reply:
The loan interest rate should be slightly higher than the interest rate of the top-non-residual tranche. Otherwise the residual tranche makes a loss or makes the same yield as the top-non-residual tranche, while instead it should get a higher yield to compensate for the higher risk.

Reply:
As a practical example, the REIF pool currently has a senior (non residual) interest rate of 5%, an average financing fee (loan interest rate) of 9.5% and a discount rate for all loans of 7% (that's probably a bit too low actually). This leads to junior (residual) getting a yield of ~15%.